File: Salary_Data_Ans.py
# For reading data set
# importing necessary libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt


# reading a csv file using pandas library
SaDa=pd.read_csv("E:\\Assignment\\4) Simple linear regression\\Salary_Data.csv")
SaDa.columns

plt.hist(SaDa.YearsExperience)
plt.boxplot(SaDa.YearsExperience,0,"rs",0)


plt.hist(SaDa.Salary)
plt.boxplot(SaDa.Salary)

plt.plot(SaDa.YearsExperience,SaDa.Salary,"bo");plt.xlabel("YearsExperience");plt.ylabel("Salary")


SaDa.Salary.corr(SaDa.YearsExperience) # # correlation value between X and Y
np.corrcoef(SaDa.Salary,SaDa.YearsExperience)

# For preparing linear regression model we need to import the statsmodels.formula.api
import statsmodels.formula.api as smf
model=smf.ols("Salary~YearsExperience",data=SaDa).fit()

# For getting coefficients of the varibles used in equation
model.params

# P-values for the variables and R-squared value for prepared model
model.summary()

model.conf_int(0.05) # 95% confidence interval

pred = model.predict(SaDa.iloc[:,0]) # Predicted values of Salary using the model

# Visualization of regresion line over the scatter plot of YearsExperience and Salary
# For visualization we need to import matplotlib.pyplot
import matplotlib.pylab as plt
plt.scatter(x=SaDa['YearsExperience'],y=SaDa['Salary'],color='red');plt.plot(SaDa['YearsExperience'],pred,color='black');plt.xlabel('YearsExperience');plt.ylabel('Salary')

pred.corr(SaDa.Salary) # 0.97

# Transforming variables for accuracy
model2 = smf.ols('Salary~np.log(YearsExperience)',data=SaDa).fit()
model2.params
model2.summary()
print(model2.conf_int(0.01)) # 99% confidence level
pred2 = model2.predict(pd.DataFrame(SaDa['YearsExperience']))
pred2.corr(SaDa.Salary)
# pred2 = model2.predict(SaDa.iloc[:,0])
pred2
plt.scatter(x=SaDa['YearsExperience'],y=SaDa['Salary'],color='green');plt.plot(SaDa['YearsExperience'],pred2,color='blue');plt.xlabel('YearsExperience');plt.ylabel('Salary')

# Exponential transformation
model3 = smf.ols('np.log(Salary)~YearsExperience',data=SaDa).fit()
model3.params
model3.summary()
print(model3.conf_int(0.01)) # 99% confidence level
pred_log = model3.predict(pd.DataFrame(SaDa['YearsExperience']))
pred_log
pred3=np.exp(pred_log)  # as we have used log(Salary) in preparing model so we need to convert it back
pred3
pred3.corr(SaDa.Salary)
plt.scatter(x=SaDa['YearsExperience'],y=SaDa['Salary'],color='green');plt.plot(SaDa.YearsExperience,np.exp(pred_log),color='blue');plt.xlabel('YearsExperience');plt.ylabel('Salary')
resid_3 = pred3-SaDa.Salary
# so we will consider the model having highest R-Squared value which is the log transformation - model3
# getting residuals of the entire data set
student_resid = model3.resid_pearson 
student_resid
plt.plot(model3.resid_pearson,'o');plt.axhline(y=0,color='green');plt.xlabel("Observation Number");plt.ylabel("Standardized Residual")

# Predicted vs actual values
plt.scatter(x=pred3,y=SaDa.Salary);plt.xlabel("Predicted");plt.ylabel("Actual")



# Quadratic model
SaDa["YearsExperience_Sq"] = SaDa.YearsExperience*SaDa.YearsExperience
model_quad = smf.ols("Salary~YearsExperience+YearsExperience_Sq",data=SaDa).fit()
model_quad.params
model_quad.summary()
pred_quad = model_quad.predict(SaDa)

model_quad.conf_int(0.05) # 
plt.scatter(SaDa.YearsExperience,SaDa.Salary,c="b");plt.plot(SaDa.YearsExperience,pred_quad,"r")

plt.scatter(np.arange(109),model_quad.resid_pearson);plt.axhline(y=0,color='red');plt.xlabel("Observation Number");plt.ylabel("Standardized Residual")

plt.hist(model_quad.resid_pearson) # histogram for residual values 

## End 

File: Salary_Data_Ans_.R
# Load Salary_Data.csv dataset
library(readr)
sa_da <- read_csv("E:\\Assignment\\4) Simple linear regression\\Salary_Data.csv")
View(sa_da)

# Exploratory data analysis
summary(sa_da)

#Scatter plot
plot(sa_da$YearsExperience, sa_da$Salary)  # plot(X,Y)

?plot

attach(sa_da)


#Correlation Coefficient (r)
cor(YearsExperience, Salary)             # cor(X,Y)

# Simple Linear Regression model
reg <- lm(Salary ~ YearsExperience) # lm(Y ~ X)

summary(reg)

pred <- predict(reg)

reg$residuals
sum(reg$residuals)

mean(reg$residuals)
sqrt(sum(reg$residuals^2)/nrow(sa_da))  #RMSE

sqrt(mean(reg$residuals^2))

confint(reg,level=0.95)
predict(reg,interval="predict")

# ggplot for adding regresion line for data
library(ggplot2)

?ggplot2

ggplot(data = sa_da, aes(x = YearsExperience, y = Salary)) + 
  geom_point(color='blue') +
  geom_line(color='red',data = sa_da, aes(x=YearsExperience, y=pred))

?ggplot2

########################
# A simple ggplot code for directly showing the line

# ggplot(sa_da,aes(YearsExperiance,Salary))+stat_summary(fun.data=mean_cl_normal) + geom_smooth(method='lm')

####################

# Logrithamic Model

# x = log(YearsExperiance); y = Salary

plot(log(YearsExperience), Salary)
cor(log(YearsExperience), Salary)

reg_log <- lm(Salary ~ log(YearsExperience))   # lm(Y ~ X)

summary(reg_log)
predict(reg_log)

reg_log$residuals
sqrt(sum(reg_log$residuals^2)/nrow(sa_da))  #RMSE

confint(reg_log,level=0.95)
predict(reg_log,interval="confidence")

######################

# Exponential Model

# x = YearsExperiance and y = log(Salary)

plot(YearsExperience, log(Salary))

cor(YearsExperience, log(Salary))

reg_exp <- lm(log(Salary) ~ YearsExperience)  #lm(log(Y) ~ X)

summary(reg_exp)

reg_exp$residuals

sqrt(mean(reg_exp$residuals^2))

logat <- predict(reg_exp)
at <- exp(logat)

error = sa_da$Salary - at
error

sqrt(sum(error^2)/nrow(sa_da))  #RMSE

confint(reg_exp,level=0.95)
predict(reg_exp,interval="confidence")

##############################
# Polynomial model with 2 degree (quadratic model)

plot(YearsExperience, Salary)
plot(YearsExperience*YearsExperience, Salary)

cor(YearsExperience*YearsExperience, Salary)

plot(YearsExperience*YearsExperience, log(Salary))

cor(YearsExperience, log(Salary))
cor(YearsExperience*YearsExperience, log(Salary))

# lm(Y ~ X + I(X*X) +...+ I(X*X*X...))

reg2degree <- lm(log(Salary) ~ YearsExperience + I(YearsExperience*YearsExperience))

summary(reg2degree)

logpol <- predict(reg2degree)
expy <- exp(logpol)

err = sa_da$Salary - expy

sqrt(sum(err^2)/nrow(sa_da))  #RMSE

confint(reg2degree,level=0.95)
predict(reg2degree,interval="confidence")

# visualization
ggplot(data = sa_da, aes(x = YearsExperience + I(YearsExperience^2), y = log(Salary))) + 
  geom_point(color='blue') +
  geom_line(color='red',data = sa_da, aes(x=YearsExperience+I(YearsExperience^2), y=logpol))


##############################
#  Polynomial model with 3 degree

reg3degree<-lm(log(Salary)~YearsExperience + I(YearsExperience*YearsExperience) + I(YearsExperience*YearsExperience*YearsExperience))

summary(reg3degree)
logpol3 <- predict(reg3degree)
expy3 <- exp(logpol3)


# visualization
ggplot(data = sa_da, aes(x = YearsExperience + I(YearsExperience^2) + I(YearsExperience^3), y = Salary)) + 
  geom_point(color='blue') +
  geom_line(color='red',data = sa_da, aes(x=YearsExperience+I(YearsExperience^2)+I(YearsExperience^3), y=expy3))

################################
